runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -11658,12 +11659,70 @@ J9::Power::CodeGenerator::inlineDirectCall(TR::Node *node, TR::Register *&result
             TR::Node *destOffset = node->getChild(2);
             TR::Node *len = node->getChild(3);
             TR::Node *byteValue = node->getChild(4);
-            dest = TR::Node::create(TR::aladd, 2, dest, destOffset);
 
-            TR::Node * copyMemNode = TR::Node::createWithSymRef(TR::arrayset, 3, 3, dest, len, byteValue, node->getSymbolReference());
-            copyMemNode->setByteCodeInfo(node->getByteCodeInfo());
+            // When using balanced GC policy with offheap allocation enabled, there are three possible cases:
+            //
+            // 1.) The object at dest is known to be a non-array object at compile time. In this scenario, the final destination address
+            //     can be calculated by simply adding dest and destOffset.
+            // 2.) The object at dest is known to be an array at compile time. In this scenario, we must load dataAddr from dest
+            //     to account for the way arrays are allocated when offheap is enabled. Once this adjustment is made, the final
+            //     destination address can be calculated by adding the adjusted dest and destOffset, similar to case (1)
+            // 3.) The type of the object at dest is unknown at compile time. In this scenario, a runtime arrayCHK must be generated to
+            //     determine whether dataAddr needs to be loaded before calculating the final destination address. Thus, dest
+            //     and destOffset will be passed to setmemoryEvaluator() separately so that both possible control flow paths are accounted
+            //     for in the generated assembly code.
+            //
+            // Note that if the default (gencon) GC policy is being used, dataAddr will never need to be loaded no matter
+            // the type of the object at dest, so all of the above cases will be treated like case (1).
+
+            bool arrayCheckNeeded, adjustmentNeeded;
+
+            if (dest->getSymbolReference() != NULL)
+            {
+               //check dstBaseAddrNode type at compile time
+               int length;
+               const char *objTypeSig = dest->getSymbolReference()->getTypeSignature(length);
+               int objSigLength = strlen("Ljava/lang/Object;");
+
+               //adjust dstBaseAddr in cases (2) and (3) (unless the object is known to be NULL)
+               adjustmentNeeded = TR::Compiler->om.isOffHeapAllocationEnabled() && comp->target().is64Bit() && !dest->isNull() &&
+                                  (objTypeSig == NULL || strncmp(objTypeSig, "Ljava/lang/Object;", objSigLength) == 0 || objTypeSig[0] == '[');
+
+               //generate arrayCHK in case (3) only
+               arrayCheckNeeded = adjustmentNeeded && (objTypeSig == NULL || strncmp(objTypeSig, "Ljava/lang/Object;", objSigLength) == 0);
+            }
+            else
+            {
+               arrayCheckNeeded = false;
```

Comment:
presumably, you intended non-offheap cases to come here, but i didn't see the code reflecting your intention.

Reply:
Rather than the non-offheap case, this else statement is there to catch the case where, for whatever reason, we can't get the symbol reference for the destination address node (i.e.: `dest->getSymbolReference() == NULL`). 

However, if that were to occur, we'd want to take the safe approach and test the object type at runtime, so the code here is indeed incorrect. Both `arrayCheckNeeded` and `adjustmentNeeded` should be set to `true` in this scenario.

Reply:
even for non-offheap cases, you went through symRef == NULL testing.  so, where is the distinction?

Reply:
One of the conditions for `adjustmentNeeded` evaluating to `true` is offheap being enabled:
```
adjustmentNeeded = TR::Compiler->om.isOffHeapAllocationEnabled() && comp->target().is64Bit() && !dest->isNull() &&
                   (objTypeUnknown || objTypeSig[0] == '[');
```

Reply:
my point was, for non-offheap cases, you don't need to go through those compile-time testings.  you can do the three-children set-up right away and alleviate the confusion, instead of squeezing the cases through the same logic by adding more and more testing conditions.